### PR TITLE
Gguf webgpu aligned tensor

### PR DESCRIPTION
**Pull request by sigma-andex**

### CODE REVIEW

File: crates/ratchet-core/src/dtype.rs [hunk 1]
```diff
@@ -46,7 +44,8 @@ impl DType {
             DType::I32 => 4,
             DType::U32 => 4,
             DType::WQ8 => 4,
-            DType::GGUF(_) => todo!(),
+            DType::GGUF(gguf::GGUFDType::Q4K(_)) => 256,
```
Comment (FL33TW00D) [inline crates/ratchet-core/src/dtype.rs:47]: So did we agree this needs to be 256?

File: crates/ratchet-core/src/storage/cpu_buffer.rs [hunk 2]
```diff
@@ -95,7 +95,7 @@ impl CPUBuffer {
     }
 
     pub fn to_slice<T: NoUninit + Pod>(&self, shape: &Shape) -> &[T] {
-        assert_eq!(self.n_bytes(), shape.numel() * std::mem::size_of::<T>());
+        // assert_eq!(self.n_bytes(), shape.numel() * std::mem::size_of::<T>());
```
Comment (FL33TW00D) [inline crates/ratchet-core/src/storage/cpu_buffer.rs:98]: Can't comment out this load bearing check, if we need an unsafe version that's cool.

File: crates/ratchet-loader/src/gguf/tensor_loader.rs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+        let (ds_segment, dmins_segment, scales_segment, qs_segment) = segments
+            .iter()
+            .collect_tuple()
+            .ok_or(anyhow!("Invalid segmentation found in Q4K"))?;
+
+        //Can't just call `to_vec` here.
+        //Do something like: https://github.com/FL33TW00D/wgpu-bench/blob/master/src/quant.rs#L105
+        // let tensor_data = tensor.to(&ratchet::Device::CPU)?.to_vec::<u32>()?;
+        // [TODO] Check if we can improve on this
+        let tensor_data = unsafe { tensor.into_bytes()? };
```
Comment (FL33TW00D) [inline crates/ratchet-loader/src/gguf/tensor_loader.rs:132]: This seems decent

File: crates/ratchet-loader/src/gguf/tensor_loader.rs [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+        let ds_alignment = ds.align_standard();
+        let mut ds_u32 = bytemuck::cast_slice::<f32, u32>(&ds).to_vec();
+        tensor_data.append(&mut ds_u32);
+
+        let inner = unsafe {
+            Tensor::from_quantized::<u32, _>(
+                &tensor_data,
+                DType::GGUF(GGUFDType::Q6K(gguf::Q6K::new())),
+                shape![tensor_blocks, 256],
```
Comment (FL33TW00D) [inline crates/ratchet-loader/src/gguf/tensor_loader.rs:221]: Do we want to hardcode 256 here?

**Comment by FL33TW00D** [comment]
Few comments, really nice work!

File: crates/ratchet-core/src/dtype.rs (hunk 1 above)
Comment (sigma-andex) [reply crates/ratchet-core/src/dtype.rs]: Fixed

File: crates/ratchet-core/src/dtype/gguf.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        let scales_len: u64 = (numel * K_SCALE_SIZE).align() as u64;
+        let offset = offset + dmins_len;
+        let scales_segment = BufferSegment::new(offset, Some(scales_len));
+
+        let qs_len: u64 = (numel * QK_K / 2).align() as u64;
+        let offset = offset + scales_len;
+        let qs_segment = BufferSegment::new(offset, Some(qs_len));
+
+        SmallVec::<[BufferSegment; 4]>::from_vec(vec![
```
[the remark this answers is not in the document]
Comment (sigma-andex) [reply crates/ratchet-core/src/dtype/gguf.rs]: Fixed

File: crates/ratchet-core/src/storage/cpu_buffer.rs (hunk 2 above)
Comment (sigma-andex) [reply crates/ratchet-core/src/storage/cpu_buffer.rs]: Fixed

**Review by FL33TW00D — APPROVED** [review]

**Pull request merged**